src/libsyntax/ext/base.rs

```diff
@@ -576,6 +576,8 @@ pub struct SyntaxExtension {
     pub kind: SyntaxExtensionKind,
     /// Span of the macro definition.
     pub span: Span,
+    /// Attributes declared on the macro definition
+    pub attrs: Vec<Attribute>,
```

Comment:
This shouldn't be a part of `SyntaxExtension`.
`SyntaxExtension` is "lowered" form of a macro definitions in which all the attributes are converted into their semantic form (see the other fields, many of them were converted from attributes).

Reply:
We store attributes [directly in the metadata](https://github.com/rust-lang/rust/blob/ef755fc7b9c4c406dae1ee5b13503c62183665fd/src/librustc_metadata/decoder.rs#L964-L975) for non-proc-macro crates - why shouldn't we do the same for proc macro crates?

Reply:
`SyntaxExtension` is not a part of metadata, it's a representation on which name resolution and expansion work.
If you look at how non-procedural macros are loaded in `fn get_macro_by_def_id`, they are loaded in "unprepared" form (`LoadedMacro::MacroDef(ast::Item)`) first, then converted (using `compile_macro`) into the "prepared" form (`SyntaxExtension`) losing all the data not necessary for resolution/expansion (like doc comments).
Procedural macros should follow a similar scheme and reuse the attribute processing functionality from `compile_macro`/`macro_rules::compile`.

Reply:
So, rustdoc is going to use the "unprepared" form, not `SyntaxExtension`, since it doesn't need to do resolution/expansion (I think).

Reply:
In that case, should I just store it as another field in [proc_macros](https://github.com/rust-lang/rust/blob/master/src/librustc_metadata/cstore.rs#L85) (and replace `(ast::Name, Lrc<SyntaxExtension>)` with a struct)?

Reply:
Yeah, somewhere "one level above" `SyntaxExtension` would work, `proc_macros` seems an appropriate place.